### DISCUSSION TITLE
Module Federation Not Tree Shaking Shared Modules Internally

### DISCUSSION THREAD

**Discussion by vgarba1** [webpack/webpack#13967] [Ideas]
I understand that Module Federation cant tree shake a shared resource on build because the module that is running the build can not know what parts of the shared resource other remote modules might need. 

However - its seems to me that Module Federation isn't even tree shaking the internals of the shared resource. 

Example:
```
// @shared/index.js
import { utilFunc } from @shared/utils/utilFunc.js

export function aTestFunction() {
  utilFunc()
}

// @shared/utils/utilFunc.js
export function utilFunc() {
   .... some code
}

export function aDifferentUtil() {
  ... some code
}
```

Say im sharing `@shared/index.js` from the example above. While we don't know exactly what the remote modules are going to use from this file, we do know that they can only use what is exported. The rest of the code can safely be tree shaken. In the example above, Module Federation doesn't need to include the entire `@shared/utils/utilFunc.js`, just any parts used by exported members of the main shared resource and it deps. In the example above, `aDifferentUtil` should be shaken out.

**Comment by nate-summercook** [comment]
Ran into the same problems, but somehow this doesn't seem to get any attention...
![gif](https://media.giphy.com/media/1Zbeweu52ZaQE/giphy.gif)

**Reply by nate-summercook** [reply]
Just found out what I did wrong here. Not sure if this is the same for you, @vgarba1, but for me it certainly was.
webpack won't tree shake the modules that you define as "shared" in your module federation configuration. Probably because it expects that you want to share this library with other parts of your code through module federation, not knowing what these will need from that dependency. If you only specify those libraries that you want to share entirely between your module federation modules, it will tree shake all the rest of your code & dependencies.
Hope this helps... I wasted ages trying to analyze what I'm doing wrong... 🫠
